### ISSUE TITLE
Messaging Extension returns with weird data

### ISSUE THREAD

**Issue by johnkellyjr** [microsoft/BotBuilder-Samples#1976]
## Sample information
1. Sample type: BotBuilder-Samples
2. Sample language: [dotnetcore ]
3. Sample name: TeamsMessagingExtensionAction (or TeamsMessagingExtensionsSearchAuthConfig)

## Describe the bug
After running the sample, a card is created on Teams. When you fill that out and submit it, it get's a card created and attached to a message in the conversation. That part seems to be fine. What happens, though, when the bot gets this message, it replies back with something like this:

{"contentType":"image/*","contentUrl":"https://smba.trafficmanager.net/amer/v3/attachments/0-cus-d5-cd9108e13a089f1ee2fae48a130ba713/views/original","content":null,"name":null,"thumbnailUrl":null}

## To Reproduce
1. Download and build the TeamsMessagingExtensionsAction, install it to Teams
2. Click on Messaging Extensions icon on bottom of Teams, a card will appear labeled "Action Messaging Extension" asking for card title, subtitle, and text. 
3. Fill in the 3 fields (card title, subtitle, and text) and click submit
4. Send it to TeamsAuth sample bot that was modified to display the incoming attachment content to the user to verify what messaging extension in Teams returns, this is handled through a "share" text command (note: if you were to send to Teams Auth sample bot something from OneDrive, it returns with proper data).
5. Instead of proper data (content type having the file name, sharepoint, ItemID, for example) it returns with:

{"contentType":"image/*","contentUrl":"https://smba.trafficmanager.net/amer/v3/attachments/0-cus-d5-365b3c14df87e8db976b51b9e62a26ed/views/original","content":null,"name":null,"thumbnailUrl":null}

## Expected behavior
We are expecting the same behavior of going to a conversation, clicking the Attach icon, selecting a file from OneDrive, and get back a similar JSON data response. It's what we currently see with the Teams Auth bot when we send it a file this way, giving it a share command as stated above.

## Screenshots
none.
![image](https://user-images.githubusercontent.com/30669826/68723679-1da33c80-056e-11ea-86a1-b53f863b16ce.png)

## Additional context
The scenario is this: Have a bot listening and waiting for shared documents. Have Messaging Extension give a list of files that are shared. The user clicks on one of the files in the list, and the bot returns back with JSON data of the file, either by a sharepoint with a file item ID or with raw JSON data with file info.

[bug]

**Comment by axelsrz** [comment]
@Virtual-Josh could you take a look at this?

**Comment by Virtual-Josh** [comment]
Hi @johnkellyjr 

It's not clear to me what you're asking. Can you clarify what your issue(s) is/are? 

Based on what you've said it sounds like there's 3 questions here. There seems to be a question about what that Action Based Messaging Extension sample currently does. 

There seems to be a question about how to get the JSON from the bot. 

There seems to be a question about how to build a bot that can actively listen for files that are shared from within Teams and do something when that happens.

**Comment by johnkellyjr** [comment]
Hi Josh,

Thank you for replying. 

The first core issue is what I'm trying to do:

- I'm trying to use a Messaging Extension attachment similar to a regular Teams "paperclip" attachment of a file from OneDrive. 
- 
- I can get a list of the user's shared files in Teams, and display them in the Messaging Extension via query (using the sample 52. Teams Messaging Extension Search Auth). 
- 
- The user selects a file, now this becomes attached in a Teams conversation. This happens in OnTeamsMessagingExtensionQueryAsync, using MessagingExtensionAttachment, creating a content type of varying trials and errors (tried hero card, thumbnail card), and also a preview (which is a thumbnail card). 

So again for this workflow:

- The user queries a file with Messaging Extension, they get the preview thumbnail card with the file name just fine. 
- 
- They click it; it appears in the conversation as an attachment just fine. 
- 
- They @TeamsAuthBot and send to another Teams Auth sample that has been slightly modified to simply echo back the JSON it was sent. 
- 
- Ideally, it would echo back the file details that I would think are sent via Teams to this sample bot. 
- 
- Instead, it echos back the {"contentType":"image/*"...etc seen in my first question.

Now, if I do the same action of attaching a file by clicking the paperclip, selecting a file from OneDrive. It gets attached to the conversation. I send that to the modified Teams Auth sample, and it returns with the file details. So this means the modified Teams Auth bot is doing what it's supposed to.


I tried this same behavior with the TeamsMessagingExtensionAction sample (and is shown in the repro steps above). There, a herocard is created, it is "submitted", it appears in the conversation. I send it to the modified Teams Auth bot which should just echo back what it got. Instead it echos back {"ContentType":"image/*" etc.

I've provided a PowerPoint mockup of what I think the flow of events is here to further help clarify. But to reiterate, I am trying to reproduce the action of creating clicking the paperclip icon, selecting the file, and then attaching a link to that file using Messaging Extension to list the shared files and getting back a sharepoint link. I realize the waters are muddied by using two different variations of the messaging extension bot (query and action) to experiment on how to do this...but the action experiment was performed as a sanity check to see what is happening.

I hope I cleared this up, but if I didn't, please let me try again.
![iamge 1](https://user-images.githubusercontent.com/30669826/68809013-d1b0d000-061f-11ea-82a0-31dcbf93035c.png)

**Comment by Virtual-Josh** [comment]
This clears part of the confusion. 

If you are trying to deal with the uploading/downloading of files I would recommend looking at our sample ([C#](https://github.com/microsoft/BotBuilder-Samples/tree/master/samples/csharp_dotnetcore/56.teams-file-upload) [JS](https://github.com/microsoft/BotBuilder-Samples/tree/master/samples/javascript_nodejs/56.teams-file-upload)) that covers that.

If you want to have the file be sent as an attachment you would need to download the file when the user selects whatever file they have. That way you can actually pass it along. 

If this doesn't solve your problem can you send over your code so I can take a look at what you're doing?

**Comment by clearab** [comment]
Hi - Let me first say that sending the results of a messaging extension with a search command to another bot isn't a supported scenario (mostly because I don't think anyone has ever tried this before 😄 ). I believe I know what is happening - The card that is rendered in the compose message area after the user selects it from the messaging extension isn't the actual card that is sent into the conversation when the user clicks send, it is just a visual representation of that card. The message payload that is getting sent to the bot is actually a pointer to that visual representation, not the actual card itself. I don't think this behavior is going to change anytime soon.

From a design perspective, why the need for two different bots? A single Teams app can contain both a bot and a messaging extension, it may be easier to process the selection on the back-end without the need to send it to another bot. You can add a tap action to the list of cards you're returning so that your web service is invoked on user selection, then process that selection w/o the need to @ the second bot. Would that possible solve your problem? If you need even more control over the process you could look at using an action command rather than a search command.

**Comment by johnkellyjr** [comment]
Hi Guys,

Thank you for offering me some different avenues to take, I am exploring them now. To answer your question, Andrew as to the design, the main reason for writing my own message extension was this: When a file is attached via a paperclip message extension from Teams Chat (the app in the left hand rail in Teams), the correct attachment details is returned. Doing the same action performed from Teams(app in the left hand rail in Teams), it always returns: {"contentType":"text/html","contentUrl":null,"content":"<span itemscope="" itemtype="http://schema.skype.com/Mention" itemid="0">Sample App(Repro)share\n","name":null,"thumbnailUrl":null}

So as I understand it, that is a pointer to the visual representation? To answer Josh's question regarding code, I am using the sample 51.teams-messaging-extensions-action as is. The modified Teams-auth bot (the second bot that I am trying to get the response from) is also nearly as is from the sample, with a modification to check for the word "share" contained in the message, and if so, to serialize the attachments and send that back as a message in LogoutDialog.cs:

private async Task<DialogTurnResult> InterruptAsync(DialogContext innerDc, CancellationToken cancellationToken = default(CancellationToken))
        {
            if (innerDc.Context.Activity.Type == ActivityTypes.Message)
            {
                var text = innerDc.Context.Activity.Text.ToLowerInvariant();

                if (text == "logout")
                {
                    // The bot adapter encapsulates the authentication processes.
                    var botAdapter = (BotFrameworkAdapter)innerDc.Context.Adapter;
                    await botAdapter.SignOutUserAsync(innerDc.Context, ConnectionName, null, cancellationToken);
                    await innerDc.Context.SendActivityAsync(MessageFactory.Text("You have been signed out."), cancellationToken);
                    return await innerDc.CancelAllDialogsAsync(cancellationToken);
                }
                else if (text.Contains("share"))
                {
                    var activity = innerDc.Context.Activity;

                    if (activity.Attachments != null)
                    {
                        foreach (Attachment file in activity.Attachments)
                        {
                            string str = JsonConvert.SerializeObject(file);
                            await innerDc.Context.SendActivityAsync(MessageFactory.Text(str), cancellationToken);
                            return await innerDc.CancelAllDialogsAsync(cancellationToken);
                        }
                    }
                }
            }

            return null;
        }

I again appreciate the prompt and useful replies, and am still looking into the #56 teams file upload sample (in the event what I was originally trying to do is impossible).

**Comment by johnkellyjr** [comment]
Hello again,

Maybe I can try a different tack. Using Bing as an example, could you please explain what is going on with the attached image?
![bing](https://user-images.githubusercontent.com/30669826/69083953-004ff180-09f8-11ea-9d40-d229d458cee2.png)

What steps are being taken to populate this list as soon as you click the Bing Messaging Extension? Is the list composed of iFrames, or are they Hero Cards? After it is clicked, what is being passed to the message box? And after it gets into the message box, how is the URL passed in as the attachment?

**Comment by Virtual-Josh** [comment]
Bing here is a bad example since. Bing is integrated into the search results as a 1st party function. Generally what you're seeing here is a search based messaging extension that has a default search. If you want to see the sample talking about this you can look [here](https://github.com/microsoft/BotBuilder-Samples/tree/master/samples/csharp_dotnetcore/50.teams-messaging-extensions-search). This sample searches Nuget (or NPM for the JS version) for our packages.

**Comment by gabog** [comment]
@johnkellyjr, it looks like Josh replied to this. Closing this issue.
